File: todoforai_edge/handlers/path_utils.py
import os
from pathlib import Path
from typing import List, Optional
from urllib.parse import urlparse, unquote
from urllib.request import url2pathname


def get_parent_directory_if_needed(path: str, root_path: Optional[str], fallback_root_paths: List[str]) -> Optional[str]:
    """
    If a relative path starts with the workspace folder name, return the parent dir of that workspace.
    """
    if os.path.isabs(path):
        return None

    all_workspace_paths = []
    if root_path:
        all_workspace_paths.append(root_path)
    if fallback_root_paths:
        all_workspace_paths.extend(fallback_root_paths)

    for workspace_path in all_workspace_paths:
        if not workspace_path:
            continue
        workspace_folder_name = os.path.basename(workspace_path.rstrip(os.sep))
        if path.startswith(workspace_folder_name + os.sep) or path == workspace_folder_name:
            workspace_parent = os.path.dirname(workspace_path)
            if workspace_parent:
                return workspace_parent
    return None


def find_file_in_workspaces(path: str, workspace_paths: List[str], primary_path: Optional[str] = None) -> Optional[str]:
    """Find a file in workspace paths, with optional primary path priority"""
    if primary_path:
        candidate_path = os.path.join(primary_path, path) if not os.path.isabs(path) else path
        candidate_path = os.path.expanduser(candidate_path)
        candidate_path = os.path.abspath(candidate_path)
        if Path(candidate_path).exists():
            return candidate_path

    for workspace_path in workspace_paths:
        candidate_path = os.path.join(workspace_path, path)
        candidate_path = os.path.expanduser(candidate_path)
        candidate_path = os.path.abspath(candidate_path)
        if Path(candidate_path).exists():
            return candidate_path

    return None


def resolve_file_path(path: str, root_path: Optional[str] = None, fallback_root_paths: List[str] = None) -> str:
    """Resolve file path using root path and fallback paths (supports file://)"""
    if isinstance(path, str) and path.startswith('file://'):
        parsed = urlparse(path)
        path = url2pathname(unquote(parsed.path))

    path = os.path.expanduser(path)

    if fallback_root_paths:
        all_paths = [root_path] + fallback_root_paths if root_path else fallback_root_paths
        parent_dir = get_parent_directory_if_needed(path, root_path, fallback_root_paths)
        if parent_dir and parent_dir not in all_paths:
            all_paths.append(parent_dir)

        found_path = find_file_in_workspaces(path, all_paths, root_path)
        if found_path:
            return found_path

    if root_path and not os.path.isabs(path):
        return os.path.join(root_path, path)

    return path

File: bun/src/path-utils.ts
import path from "path";
import fs from "fs";
import { fileURLToPath } from "url";

export class WorkspacePathNotFoundError extends Error {
  missingRoots: string[];
  constructor(filePath: string, missingRoots: string[]) {
    super(`File not found: ${filePath} — workspace path(s) do not exist: ${missingRoots.join(", ")}`);
    this.name = "WorkspacePathNotFoundError";
    this.missingRoots = missingRoots;
  }
}

function expandUser(p: string): string {
  if (p.startsWith("~")) return p.replace("~", process.env.HOME || "~");
  return p;
}

function getParentDirectoryIfNeeded(filePath: string, rootPath: string | undefined, fallbackRootPaths: string[]): string | null {
  if (path.isAbsolute(filePath)) return null;

  const allPaths: string[] = [];
  if (rootPath) allPaths.push(rootPath);
  allPaths.push(...fallbackRootPaths);

  for (const wp of allPaths) {
    if (!wp) continue;
    const folderName = path.basename(wp.replace(/[/\\]+$/, ""));
    if (filePath.startsWith(folderName + path.sep) || filePath === folderName) {
      const parent = path.dirname(wp);
      if (parent) return parent;
    }
  }
  return null;
}

function findFileInWorkspaces(filePath: string, workspacePaths: string[], primaryPath?: string): string | null {
  if (primaryPath) {
    const candidate = path.isAbsolute(filePath) ? filePath : path.resolve(expandUser(primaryPath), filePath);
    if (fs.existsSync(candidate)) return candidate;
  }
  for (const wp of workspacePaths) {
    const candidate = path.resolve(expandUser(wp), filePath);
    if (fs.existsSync(candidate)) return candidate;
  }
  return null;
}

function checkRootsExist(filePath: string, rootPath?: string, fallbackRootPaths?: string[]): void {
  const roots: string[] = [];
  if (rootPath) roots.push(rootPath);
  if (fallbackRootPaths) roots.push(...fallbackRootPaths);
  const missing = roots.filter(r => r && !fs.existsSync(r));
  if (missing.length) throw new WorkspacePathNotFoundError(filePath, missing);
}

export function resolveFilePath(filePath: string, rootPath?: string, fallbackRootPaths: string[] = []): string {
  // Handle file:// URLs
  if (filePath.startsWith("file://")) {
    filePath = fileURLToPath(filePath);
  }
  filePath = expandUser(filePath);

  checkRootsExist(filePath, rootPath, fallbackRootPaths);

  if (fallbackRootPaths.length) {
    const allPaths = rootPath ? [rootPath, ...fallbackRootPaths] : [...fallbackRootPaths];
    const parentDir = getParentDirectoryIfNeeded(filePath, rootPath, fallbackRootPaths);
    if (parentDir && !allPaths.includes(parentDir)) allPaths.push(parentDir);

    const found = findFileInWorkspaces(filePath, allPaths, rootPath);
    if (found) return found;
  }

  if (rootPath && !path.isAbsolute(filePath)) {
    return path.join(rootPath, filePath);
  }

  return filePath;
}

export function getPlatformDefaultDirectory(): string {
  try {
    const home = process.env.HOME || process.env.USERPROFILE;
    if (home && fs.existsSync(home)) return path.resolve(home);
  } catch {}
  return process.cwd();
}

export function getPathOrDefault(p?: string): string {
  if (!p || p === "." || p === "") return getPlatformDefaultDirectory();
  return p;
}
